### PR TITLE
Fix JSON.stringify return type (#18879), reduce usage of any in JSON.parse

### LINKED ISSUE #18879: Type signature for JSON.stringify does not include undefined in the return type
Although calling `JSON.stringify` may return undefined (as documented here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#Description), it is not included in the type signature provided with TypeScript.

### PR DESCRIPTION
**Pull request by clarfonthey**
Fixes #18879.

This reflects the types of the actual `JSON.stringify` function, which may return `string | undefined`. This approach chooses to add extra overloads specific to the JSON-valid types (specifically `boolean | number | string | object`) to allow the caller to ensure that the value is definitely a string. To make these overloads effective, the input that may potentially return `undefined` is marked with `unknown`, not `any`.

This also modifies the replacer to `JSON.parse` to follow the same format as the new replace for `stringify`, which now accepts `unknown` instead of `any`. To minimise the cascading of changes made, however, `parse` still returns `any` instead of `unknown`.

### CODE REVIEW

**Comment by clarfonthey** [comment]
I'm going to try one more time, but I genuinely can't reproduce the errors happening on CI, locally. I rebased on main multiple times and ran multiple variations of the test commands, including literally copying the same stuff CI does, and it just did not result in the same errors locally. Everything passes on my end. `node --version` is 18.7.0 and it doesn't matter whether I do a regular `npm install` or `npm ci`; it just doesn't work.

File: src/lib/es5.d.ts [hunk 1]
```diff
@@ -1113,21 +1113,35 @@ interface JSON {
      * @param reviver A function that transforms the results. This function is called for each member of the object.
      * If a member contains nested objects, the nested objects are transformed before the parent object is.
      */
-    parse(text: string, reviver?: (this: any, key: string, value: any) => any): any;
+    parse(text: string, reviver?: (this: unknown, key: string, value: unknown) => unknown): any;
     /**
-     * Converts a JavaScript value to a JavaScript Object Notation (JSON) string.
-     * @param value A JavaScript value, usually an object or array, to be converted.
+     * Converts a JSON value to a JavaScript Object Notation (JSON) string.
```
Comment (Josh-Cena) [inline src/lib/es5.d.ts:1118]: There's no such thing as a _JSON value_. JSON is a _notation_, or a text format. There's only _JSON-serializable value_.

Comment (clarfonthey) [reply src/lib/es5.d.ts:1118]: Sure, I'll change it to JSON-representable to match the official ECMAScript spec.

**Comment by clarfonthey** [comment]
Note to any TypeScript maintainers reading this: the fact that baselines hard-code the line and column of items in the library source code makes this very difficult to properly maintain. Given how changes seem to be merged daily that change these files, this makes library improvements additionally painful, even if it's for a very important change.

I don't blame the actual team members responsible for reviewing PRs, but would like to specifically call out Microsoft, a several-billion-dollar company, for not allocating the resources to the TS team to adequately review PRs to what has become a critical piece of infrastructure for many companies. This is not a volunteer project, but something being developed by a for-profit corporation with control of much of the Node ecosystem.

I'm willing to continue to put up with merging in the main branch for another week or so, but beyond that, I'm going to just give up and face the fact that this isn't going to be merged. I'm almost certain this is the reason why this bug, which has had multiple fix PRs opened since it was initially reported, has never been fixed-- the project changes rapidly, is fragile to these sorts of changes, and there's inadequate staff allocated to actually helping contributors merge changes in a manner that reflects that.

Doubt anyone's actually going to be able to take action on this, but figured I'd post a comment for those burned by this bug to read and sympathise with. Hopefully, this doesn't matter too much and this actually does get merged.

**Comment by RyanCavanaugh** [comment]
The CI runs a second pass to ensure that TypeScript can still build itself after any proposed change. You introduced build errors into `shims.ts`, as shown on the PR page itself:

image

You can reproduce this locally by running
```
gulp lkg
gulp local
```
I haven't seen a PR break the build in this particular way before, but we should probably update the CONTRIBUTING.md instructions to address this edge case.

**Comment by clarfonthey** [comment]
So I should add, this particular case came up after the latest change, but before I was getting other baseline diffs showing up that I could not actually get to generate locally, no matter how hard I tried; it just said that the files were generated. This change is actually very clear in the diffs that run in CI.

I'll see if I can add these steps to `CONTRIBUTING.md`. Thank you for helping out. ❤️

File: src/lib/es5.d.ts [hunk 2]
```diff
@@ -1534,11 +1534,11 @@ interface Promise<T> {
...
-        T extends object & { then(onfulfilled: infer F, ...args: infer _): any } ? // `await` only unwraps object types with a callable `then`. Non-object types are not unwrapped
-            F extends ((value: infer V, ...args: infer _) => any) ? // if the argument to `then` is callable, extracts the first argument
-                Awaited<V> : // recursively unwrap the value
-                never : // the argument to `then` was not callable
-        T; // non-object or non-thenable
+    T extends object & { then(onfulfilled: infer F, ...args: infer _): any } ? // `await` only unwraps object types with a callable `then`. Non-object types are not unwrapped
+    F extends ((value: infer V, ...args: infer _) => any) ? // if the argument to `then` is callable, extracts the first argument
+    Awaited<V> : // recursively unwrap the value
+    never : // the argument to `then` was not callable
+    T; // non-object or non-thenable
```
Comment (clarfonthey) [inline src/lib/es5.d.ts:1541]: This is how typescript-language-server auto-formats this file -- I don't prefer it, but the best way I have at the moment to make the changes not look horrendous as I'm modifying them is to format-on-save and then run eslint --fix.

**Comment by clarfonthey** [comment]
So I don't understand why CI is passing now -- I didn't finish fixing all the errors last night and they're still showing up for me locally, so, I'm planning to do more work later on it.

File: src/lib/es5.d.ts (hunk 2 above)
Comment (jakebailey) [reply src/lib/es5.d.ts:1541]: I would recommend disabling formatting altogether on this repo; we don't use any particular style besides the lines, so it's just not a good time for those who use format on save (which I normally do, except here). I'm hoping that can change one day, but checker is just a huge beast.

File: src/harness/harnessUtils.ts [hunk 3]
```diff
@@ -148,7 +148,7 @@ namespace Utils {
         }
     }
 
-    function isNodeOrArray(a: any): boolean {
+    function isNodeOrArray(a: any): a is ts.Node {
```
Comment (jakebailey) [inline src/harness/harnessUtils.ts:151]: Is this accurate? I'd think this should mention NodeArray, which is not a Node.

File: src/testRunner/unittests/tsbuildWatch/projectsBuilding.ts [hunk 4]
```diff
@@ -11,29 +11,31 @@ namespace ts.tscWatch {
             return { path: `./pkg${index}` };
         }
         function pkgFiles(index: number): File[] {
+            const tsconfig = {
+                compilerOptions: { composite: true },
```
Comment (jakebailey) [inline src/testRunner/unittests/tsbuildWatch/projectsBuilding.ts:15]: Were these changes necessary to make the code compile?

Comment (clarfonthey) [reply src/testRunner/unittests/tsbuildWatch/projectsBuilding.ts:15]: Yes.

File: src/harness/harnessUtils.ts (hunk 3 above)
Comment (clarfonthey) [reply src/harness/harnessUtils.ts:151]: Will double-check.

File: src/harness/fourslashImpl.ts [hunk 5]
```diff
@@ -1341,7 +1341,7 @@ namespace FourSlash {
             }));
         }
 
-        public verifyQuickInfoAt(markerName: string | Range, expectedText: string, expectedDocumentation?: string, expectedTags?: {name: string; text: string;}[]) {
+        public verifyQuickInfoAt(markerName: string | Range, expectedText: string, expectedDocumentation?: string, expectedTags?: { name: string; text: string; }[]) {
```
Comment (sandersn) [inline src/harness/fourslashImpl.ts:1344]: please undo these formatting changes. We'll create a giant formatting PR when we (someday, maybe) start using a formatter.

File: src/harness/fourslashImpl.ts [hunk 6]
```diff
@@ -4504,8 +4504,8 @@ namespace FourSlash {
         };
     }
 
-    function stringify(data: any, replacer?: (key: string, value: any) => any): string {
-        return JSON.stringify(data, replacer, 2);
+    function stringify(data: unknown, replacer?: (key: string, value: unknown) => unknown): string {
+        return JSON.stringify(data, replacer, 2) ?? "undefined";
```
Comment (sandersn) [inline src/harness/fourslashImpl.ts:4508]: this would have been equivalent to `... ?? ""` before, right? Do any baselines change?

File: src/lib/es5.d.ts [hunk 7]
```diff
@@ -1111,21 +1111,35 @@ interface JSON {
...
      */
-    stringify(value: any, replacer?: (number | string)[] | null, space?: string | number): string;
+    stringify(value: boolean | number | string | object, replacer?: (number | string)[] | null, space?: string | number): string;
+    /**
+     * Converts a JavaScript value to a JavaScript Object Notation (JSON) string or undefined.
+     * @param value A JavaScript value to be converted, which could potentially be a non-JSON-representable value.
+     * @param replacer A function that transforms the results.
+     * @param space Adds indentation, white space, and line break characters to the return-value JSON text to make it easier to read.
+     */
+    stringify(value: unknown, replacer?: (this: unknown, key: string, value: unknown) => unknown, space?: string | number): string | undefined;
```
Comment (sandersn) [inline src/lib/es5.d.ts:1135]: `this: unknown` requires `this` to be `unknown` or a supertype of unknown if it's specified. That doesn't sound right to me. What types are supposed to allowed and disallowed for `this` here?

File: src/testRunner/unittests/tsbuildWatch/projectsBuilding.ts (hunk 4 above)
Comment (sandersn) [reply src/testRunner/unittests/tsbuildWatch/projectsBuilding.ts:15]: is it because of contextual typing? or maybe object literal freshness?

**Review by sandersn — CHANGES_REQUESTED** [review]
1. this parameter type seems wrong
2. I am curious about the number of objects that had to be extracted to a variable.
3. I would like the whitespace changes to be reverted.

File: src/harness/fourslashImpl.ts (hunk 5 above)
Comment (clarfonthey) [reply src/harness/fourslashImpl.ts:1344]: Okay. Keep in mind that this is just the formatting the langserver did automatically, and not a special formatter.

File: src/harness/fourslashImpl.ts (hunk 6 above)
Comment (clarfonthey) [reply src/harness/fourslashImpl.ts:4508]: Actually, the previous code would have returned `undefined` (not string), which is why I was conservative and returned the string `"undefined"` since that's what would happen when it's cast to a string.

File: src/lib/es5.d.ts (hunk 7 above)
Comment (clarfonthey) [reply src/lib/es5.d.ts:1135]: `this` is always set to the value being replaced, which doesn't have any distinct type. It could be a symbol or function or some other value added to the standard in the distant future, which is why I was conservative.

When you say "supertype of `unknown`"... I thought that `unknown` was the maximal type?

**Comment by clarfonthey** [comment]
I've been mostly unable to work on this this week due to real-life moving hell, but was planning to continue this either this week or next.

In terms of the object literal extraction, to me understanding, it's because the type checking literally differs between lvalues and rvalues, or literals and bindings, or whatever pair of terms you want to use. It prefers to treat the value as `unknown` rather than directly inferring `object`.

I chose specifically `object` instead of some other type like a `Record` since tsc infers this directly when checking `typeof`, and it's the type that most closely matches the spec. But it's unhappy using this type when directly comparing object literals for some unknown reason. If you can figure out the cause of this, I'm fine undoing my changes.

I was testing via `npx gulp lkg && npx gulp tests`; I didn't want to run in parallel since I wanted to do other things while it was running.

File: src/harness/fourslashImpl.ts (hunk 5 above)
Comment (clarfonthey) [reply src/harness/fourslashImpl.ts:1344]: In the final PR I can remove it, I was just tired of constantly removing it when it was done automatically, when my main concern was getting tests to pass.

**Comment by clarfonthey** [comment]
Yeah I, lost all desire to work on this, so, I'm just going to close this PR so people don't think I'm going to try this again.
